### PR TITLE
Analyzer binary dependent environment

### PR DESCRIPTION
**Pull request by dkrupp**
If a pre-package analyzer binary (e.g. clang) is found first, the LD_LIBRARY_PATH must be extended. On the other hand, if the analyzer binary is taken from the users machine, the original environment must be used to execute the binary.

This patch introduces an analyzer binary dependent environment initialization.

### CODE REVIEW

File: analyzer/codechecker_analyzer/analyzers/analyzer_types.py [hunk 1]
```diff
@@ -104,7 +104,8 @@ def is_ignore_conflict_supported():
     proc = subprocess.Popen([context.replacer_binary, '--help'],
                             stdout=subprocess.PIPE,
                             stderr=subprocess.PIPE,
-                            env=context.analyzer_env,
+                            env=context
+                            .get_analyzer_env("clang-apply-replacements"),
```
Comment (bruntib) [inline analyzer/codechecker_analyzer/analyzers/analyzer_types.py:108]:
```suggestion
                            .get_analyzer_env(
                                os.path.basename(context.replacer_binary)),
```

File: analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py [hunk 2]
```diff
@@ -51,7 +51,7 @@ def parse_clang_help_page(
         help_page = subprocess.check_output(
             command,
             stderr=subprocess.STDOUT,
-            env=analyzer_context.get_context().analyzer_env,
+            env=analyzer_context.get_context().get_analyzer_env("clangsa"),
```
Comment (bruntib) [inline analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py:54]:
```suggestion
            env=analyzer_context.get_context()\
            .get_analyzer_env(ClangSA.ANALYZER_NAME),
```

File: analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py [hunk 3]
```diff
@@ -207,7 +207,7 @@ def ctu_capability(cls):
         if not cls.__ctu_autodetection:
             cls.__ctu_autodetection = CTUAutodetection(
                 cls.analyzer_binary(),
-                analyzer_context.get_context().analyzer_env)
+                analyzer_context.get_context().get_analyzer_env("clangsa"))
```
Comment (bruntib) [inline analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py:210]:
```suggestion
                analyzer_context.get_context()\
                .get_analyzer_env(ClangSA.ANALYZER_NAME))
```

File: analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py [hunk 4]
```diff
@@ -606,7 +606,7 @@ def construct_result_handler(self, buildaction, report_output,
     def construct_config_handler(cls, args):
 
         context = analyzer_context.get_context()
-        environ = context.analyzer_env
+        environ = context.get_analyzer_env("clangsa")
```
Comment (bruntib) [inline analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py:609]:
```suggestion
        environ = context.get_analyzer_env(ClangSA.ANALYZER_NAME)
```

File: analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py [hunk 5]
```diff
@@ -270,7 +270,8 @@ def get_analyzer_checkers(cls):
             if cls.__analyzer_checkers:
                 return cls.__analyzer_checkers
 
-            environ = analyzer_context.get_context().analyzer_env
+            environ = analyzer_context\
+                .get_context().get_analyzer_env("clang-tidy")
```
Comment (bruntib) [inline analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py:274]:
```suggestion
                .get_context().get_analyzer_env(cls.ANALYZER_NAME)
```

File: analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py [hunk 6]
```diff
@@ -297,7 +298,8 @@ def get_checker_config(cls):
         try:
             result = subprocess.check_output(
                 [cls.analyzer_binary(), "-dump-config", "-checks=*"],
-                env=analyzer_context.get_context().analyzer_env,
+                env=analyzer_context.get_context()
+                .get_analyzer_env("clang-tidy"),
```
Comment (bruntib) [inline analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py:302]:
```suggestion
                .get_analyzer_env(cls.ANALYZER_NAME),
```

File: analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py [hunk 7]
```diff
@@ -313,7 +315,8 @@ def get_analyzer_config(cls):
...
```
Comment (bruntib) [inline analyzer/codechecker_analyzer/analyzers/clangtidy/analyzer.py:319]:
```suggestion
                .get_analyzer_env(cls.ANALYZER_NAME),
```

File: analyzer/codechecker_analyzer/analyzers/cppcheck/analyzer.py [hunk 8]
```diff
@@ -393,7 +393,7 @@ def construct_config_handler(cls, args):
             # No cppcheck arguments file was given in the command line.
             LOG.debug_analyzer(aerr)
 
-        check_env = context.analyzer_env
+        check_env = context.get_analyzer_env("cppcheck")
```
Comment (bruntib) [inline analyzer/codechecker_analyzer/analyzers/cppcheck/analyzer.py:396]:
```suggestion
        check_env = context.get_analyzer_env(cls.ANALYZER_NAME)
```

**Review by bruntib — CHANGES_REQUESTED** [review]
`SourceAnalyzer` class has a virtual function: `analyze()`. This sets the environment accordingly. The inherited `Gcc` and `Cppcheck` classes override this function which also set the environment. I think, we shouldn't expect the child classes do this job. The `analyze()` functions in these two inherited classes could be removed.

**Comment by dkrupp** [comment]
> `SourceAnalyzer` class has a virtual function: `analyze()`. This sets the environment accordingly. The inherited `Gcc` and `Cppcheck` classes override this function which also set the environment. I think, we shouldn't expect the child classes do this job. The `analyze()` functions in these two inherited classes could be removed.

Good point! Fixed!

**Comment by dkrupp** [comment]
Looks like this PR changed the behaviour in CodeChecker:

https://github.com/Ericsson/codechecker/pull/3919/files

the ld_logger.so binary is selected based on the host OS platform instead of the executed binary...

**Review by bruntib — APPROVED** [review]

**Pull request merged**